backend/engine/plugins/ghas_secrets/main.py

```diff
@@ -0,0 +1,230 @@
+
+    return results, event_info
+
+
+def _normalize_secret_type(secret_type: str) -> str:
+    # There is a large number of secret types supported by GHAS. Some of them need to be normalized
+    # to be aligned with the types returned by other Artemis plugins.
+    #
+    # https://docs.github.com/en/code-security/secret-scanning/secret-scanning-patterns
+    if secret_type.startswith("aws"):
+        return "aws"
+    elif secret_type.startswith("google"):
+        return "google"
+    elif "_ssh_" in secret_type:
+        return "ssh"
+    elif secret_type in ["slack_incoming_webhook_url", "slack_workflow_webhook_url"]:
+        return "slack"
+    return secret_type  # Keep the original type as a fallthrough
+
+
+def _get_alerts(gh: GITHUB_INFO) -> list[dict]:
+    alerts = _github_api_get(gh, "secret-scanning/alerts", query={"state": "open"}, paged=True)
+    LOG.info("Retrieved %s alerts", len(alerts))
+    return alerts
+
+
+def _get_locations(gh: GITHUB_INFO, alert_id: str) -> list[dict]:
+    locations = _github_api_get(gh, f"secret-scanning/alerts/{alert_id}/locations", paged=True)
+    LOG.debug("Retrieved %s location(s) for alert %s", len(locations), alert_id)
+    return locations
+
+
+def _validate_location(location: dict, path: str) -> Tuple[bool, str, str]:
+    valid = False
+    author = None
+    author_timestamp = None
+
+    r = subprocess.run(
+        ["git", "merge-base", "--is-ancestor", location["details"]["commit_sha"], "HEAD"], cwd=path, capture_output=True
+    )
+    if r.returncode == 0:
+        # The commit with the secret is in the current branch
+        valid = True
+        r = subprocess.run(
+            [
+                "git",
+                "show",
+                "--format=%an <%ae>%n%aI",  # Author Name <Author Email>\nISO 8601 Author Date
+                "--no-patch",  # Don't care about the diff
+                location["details"]["commit_sha"],
+            ],
+            cwd=path,
+            capture_output=True,
+        )
+        if r.returncode == 0:
+            author, author_timestamp = r.stdout.decode("utf-8").strip().split("\n")
+
+    return valid, author, author_timestamp
+
+
+def _get_authorization(org: str, github_secret: str) -> str:
+    # Attempt to get an app installation token for the organization
+    github_app = GithubApp()
+    token = github_app.get_installation_token(org)
+    if token is not None:
+        return f"token {token}"
+
+    # Fall back to getting the PAT
+    key = _get_api_key(github_secret)
+    return f"bearer {key}"
+
+
+def _get_api_key(service_secret):
+    from artemislib.aws import AWSConnect  # pylint: disable=import-outside-toplevel
+
+    aws_connect = AWSConnect()
+    secret = aws_connect.get_secret(f"{APPLICATION}/{service_secret}")
+    if secret:
+        return secret.get("key")
+    return None
+
+
+def _github_api_get(gh: GITHUB_INFO, path: str = None, query: dict = None, paged=False):
```

Comment:
Is there a way this can be factored out so that it is usable by other plugins (for example, making it a part of `GithubAPI` in `artemislib`)?

At minimum, I think the header-setting stuff would be good factored out into a common library. The paging stuff would also be useful

Reply:
Updated